test/infrastructure/docker/e2e/docker_test.go

```diff
@@ -0,0 +1,202 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"fmt"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/cluster-api/bootstrap/kubeadm/types/v1beta1"
+	"sigs.k8s.io/cluster-api/test/framework"
+
+	corev1 "sigs.k8s.io/cluster-api/api/v1alpha3"
+	bootstrapv1 "sigs.k8s.io/cluster-api/bootstrap/kubeadm/api/v1alpha2"
+	infrav1 "sigs.k8s.io/cluster-api/test/infrastructure/docker/api/v1alpha2"
+)
+
+var _ = Describe("Docker", func() {
+	Describe("Cluster Creation", func() {
+
+		namespace := "default"
+
+		clusterGen := &ClusterGenerator{}
+		nodeGen := &NodeGenerator{}
+
+		Context("One node cluster", func() {
+			It("should create a single node cluster", func() {
+				cluster, infraCluster := clusterGen.GenerateCluster(namespace)
+				singleNode := nodeGen.GenerateNode(cluster.GetName())
+				input := &framework.OneNodeClusterInput{
+					Management:    mgmt,
+					Cluster:       cluster,
+					InfraCluster:  infraCluster,
+					Node:          singleNode,
+					CreateTimeout: 3 * time.Minute,
+				}
+
+				framework.OneNodeCluster(input)
+
+				cleanupInput := &framework.CleanUpInput{
+					Management: mgmt,
+					Cluster:    cluster,
+				}
+
+				framework.CleanUp(cleanupInput)
+			})
+		})
+
+		Context("Multi-node controlplane cluster", func() {
+			It("should create a multi-node controlplane cluster", func() {
+				cluster, infraCluster := clusterGen.GenerateCluster(namespace)
+				nodes := make([]framework.Node, 3)
+				for i := range nodes {
+					nodes[i] = nodeGen.GenerateNode(cluster.Name)
+				}
+
+				input := &framework.MultiNodeControlplaneClusterInput{
+					Management:        mgmt,
+					Cluster:           cluster,
+					InfraCluster:      infraCluster,
+					ControlplaneNodes: nodes,
+					CreateTimeout:     2 * time.Minute,
+				}
+				framework.MultiNodeControlPlaneCluster(input)
+
+				cleanupInput := &framework.CleanUpInput{
+					Management: mgmt,
+					Cluster:    cluster,
+				}
+
+				framework.CleanUp(cleanupInput)
+			})
+		})
+	})
+})
+
+type ClusterGenerator struct {
+	counter int
+}
+
+func (c *ClusterGenerator) GenerateCluster(namespace string) (*corev1.Cluster, *infrav1.DockerCluster) {
```

Comment:
Longer term, I'd like to see us offer a subset of this as part of the framework, where the user defines the infra resource and some parameterized input for the common config, and we do handle the linking of the resources and general housekeeping for consumers.

Reply:
Yeah, agreed. I'd like to see the current abstraction live for a bit and see how painful it is and get some feedback after usage. But I agree that there's no way this will live for very long